File: service/simple-js/buyer.js
var autobahn = require("autobahn");
var xbr = require("autobahn-xbr");
var web3 = require("web3");
var BN = web3.utils.BN;

console.log("Running on Autobahn " + autobahn.version);
//Private key of your buyer delegate
const BUYER_PRIVATEKEY = "0x16A60CED74B0D0523567B2261CF3E1C41048036D291D7132BF353E10B74F817A";
//Market maker address
const MARKET_MAKER_ADDR = "0x3E5e9111Ae8eB78Fe1CC3bb8915d5D461F3Ef9A9";
//TEAM_ID
const TEAM_ID = "team1";
//TICKET
const TICKET = "apple pear orange fig";
//TOPIC TO SUBSCRIBE TO
const TOPIC = "com.conti.hackathon.team1.topic3";

//Connect to Crossbarfx node
var connection = new autobahn.Connection({
    realm: "realm1",
    authmethods: ["ticket"],
    authid: TEAM_ID,
    onchallenge: onchallenge,
    transports: [{
        url: "wss://continental2.crossbario.com/ws",
        type: "websocket",
        serializers: [new autobahn.serializer.CBORSerializer()]
    }]
});

//Only for authentication on crossbar
function onchallenge(session, method, extra) {
    if (method === "ticket") {
        return TICKET;
    } else {
        throw "don't know how to authenticate using '" + method + "'";
    }
};

// callback fired upon new WAMP session
connection.onopen = function(session, details) {
    console.log("WAMP session connected:", details);

    // the XBR token has 18 decimals
    const decimals = new BN("1000000000000000000");

    // maximum price we are willing to pay per (single) key: 100 XBR
    const max_price = new BN("100").mul(decimals);

    // instantiate a simple buyer
    var buyer = new xbr.SimpleBuyer(
        MARKET_MAKER_ADDR,
        BUYER_PRIVATEKEY,
        max_price
    );

    // start buying ..
    buyer.start(session).then(
        // success: we've got an active payment channel with remaining balance ..
        function(balance) {
            console.log(
                "Buyer has started, remaining balance in active payment channel is " +
                balance.div(decimals) +
                " XBR"
            );

            session.subscribe(TOPIC, function(args, kwargs, details) {
                let [key_id, enc_ser, ciphertext] = args;

                // decrypt the XBR payload, potentially automatically buying a new data encryption key
                buyer.unwrap(key_id, enc_ser, ciphertext).then(
                    function(payload) {
                        console.log("Received event " + details.publication, payload);
                    },
                    function(failure) {
                        console.log(failure);
                        process.exit(1);
                    }
                );
            }, { match: "prefix" });
        },
        // we don't have an active payment channel => bail out
        function(error) {
            console.log("Failed to start buyer:", error);
            process.exit(1);
        }
    );
};

// open WAMP session
connection.open();

File: service/simple-python/buyer.py
#!/usr/bin/env python3
"""Simple Buyer example.
"""

from autobahn.twisted.xbr import SimpleBuyer
from autobahn.twisted.component import Component, run
from autobahn.wamp.types import SubscribeOptions
import binascii
import os

import txaio
txaio.use_twisted()


DEFAULT_AUTHID = 'team1'
DEFAULT_TICKET = 'apple pear orange fig'

comp = Component(
    transports=os.environ.get(
        'XBR_INSTANCE', 'wss://continental2.crossbario.com/ws'),
    realm=os.environ.get('XBR_REALM', 'realm1'),
    authentication={
        'ticket': {
            'authid': os.environ.get('XBR_AUTHID', DEFAULT_AUTHID),
            'ticket': os.environ.get('XBR_TICKET', DEFAULT_TICKET),
        },
    },
    extra={
        'market_maker_adr': os.environ.get('XBR_MARKET_MAKER_ADR', '0x3e5e9111ae8eb78fe1cc3bb8915d5d461f3ef9a9'),
        'buyer_privkey': os.environ.get('XBR_BUYER_PRIVKEY', '0x16A60CED74B0D0523567B2261CF3E1C41048036D291D7132BF353E10B74F817A'),
    }
)


@comp.on_join
async def joined(session, details):
    print('Buyer session joined', details)

    market_maker_adr = binascii.a2b_hex(
        session.config.extra['market_maker_adr'][2:])
    print('Using market maker adr:', session.config.extra['market_maker_adr'])

    buyer_privkey = binascii.a2b_hex(session.config.extra['buyer_privkey'][2:])

    max_price = 100 * 10 ** 18
    buyer = SimpleBuyer(market_maker_adr, buyer_privkey, max_price)
    balance = await buyer.start(session, details.authid)
    balance = int(balance / 10 ** 18)
    print("Remaining balance: {} XBR".format(balance))

    async def on_event(key_id, enc_ser, ciphertext, details=None):
        try:
            payload = await buyer.unwrap(key_id, enc_ser, ciphertext)
        except Exception as e:
            print(e)
            session.leave()
        else:
            print('Received event {}:'.format(details.publication), payload)

    await session.subscribe(on_event, "com.conti.hackathon.team1", options=SubscribeOptions(details=True))


if __name__ == '__main__':
    run([comp])
